cornac/models/trirank/recom_trirank.py

```diff
@@ -0,0 +1,326 @@
+        if self.verbose:
+            print("Building symmetric normalized matrices R, X, Y")
+        self.R = self._symmetrical_normalization(train_set.csr_matrix)
+        self.X = self._symmetrical_normalization(X)
+        self.Y = self._symmetrical_normalization(Y)
+
+        if self.verbose:
+            total_time = time() - start_time
+            print("Building matrices completed in %d s" % total_time)
+
+    def fit(self, train_set, val_set=None):
+        """Fit the model to observations.
+
+        Parameters
+        ----------
+        train_set: :obj:`cornac.data.Dataset`, required
+            User-Item preference data as well as additional modalities.
+
+        val_set: :obj:`cornac.data.Dataset`, optional, default: None
+            User-Item preference data for model selection purposes (e.g., early stopping).
+
+        Returns
+        -------
+        self : object
+        """
+        Recommender.fit(self, train_set, val_set)
+        self._init()
+
+        if not self.trainable:
+            return self
+
+        # Offline training: Build item-aspect matrix X and user-aspect matrix Y
+        self._create_matrices(train_set)
+        return self
+
+    def _online_recommendation(self, user):
+        # Algorithm 1: Online recommendation line 5
+        p_0 = self.train_set.csr_matrix[user]
+        indices = p_0.nonzero()
+        p_0[indices] = 1
+        p_0 = p_0.toarray().squeeze()
+        a_0 = self.Y[user].toarray().squeeze()
+        u_0 = np.zeros(self.train_set.csr_matrix.shape[0])
+        u_0[user] = 1
+
+        # Algorithm 1: Online training line 6
+        if p_0.any():
+            p_0 /= np.linalg.norm(p_0, 1)
+        if a_0.any():
+            a_0 /= np.linalg.norm(a_0, 1)
+        if u_0.any():
+            u_0 /= np.linalg.norm(u_0, 1)
+
+        # Algorithm 1: Online recommendation line 7
+        p = self.p.copy()
+        a = self.a.copy()
+        u = self.u.copy()
+
+        # Algorithm 1: Online recommendation line 8
+        prev_p = p
+        prev_a = a
+        prev_u = u
+        while True:
+            # eq. 4
+            u_denominator = self.alpha + self.gamma + self.eta_U + EPS
+            u = (
+                self.alpha / u_denominator * self.R * p
+                + self.gamma / u_denominator * self.Y * a
+                + self.eta_U / u_denominator * u_0
+            ).squeeze()
+            p_denominator = self.alpha + self.beta + self.eta_P + EPS
+            p = (
+                self.alpha / p_denominator * self.R.T * u
+                + self.beta / p_denominator * self.X * a
+                + self.eta_P / p_denominator * p_0
+            ).squeeze()
+            a_denominator = self.gamma + self.beta + self.eta_A + EPS
+            a = (
+                self.gamma / a_denominator * self.Y.T * u
+                + self.beta / a_denominator * self.X.T * p
+                + self.eta_P / a_denominator * a_0
+            ).squeeze()
+
+            if (
```

Comment:
Should we also have an option of `max_iter` for breaking the loop? As discussed in the paper that "For convergence, one can monitor Q(f)’s value until it stabilizes or set a maximum number of iterations."